[Clang][inlineasm] Add special support for "rm" output constraints

bwendling opened the pull request:
Clang isn't able to support multiple constraints on inputs and outputs,
like "rm". Instead, it picks the "safest" one to use, i.e. the memory
constraint for "rm". This leads to obviously horrible code:

  asm __volatile__ ("pushf\n\t"
                    "popq %0"
                    : "=rm" (x));

is compiled to:

        pushf
	popq -8(%rsp)
	movq	-8(%rsp), %rax

It gets worse when inlined into other functions, because it may introduce
a stack where none is needed.

With this change, Clang now generates IR for the more optimistic choice
("r"). All but the fast register allocator are able to fold registers if
it turns out that register pressure is too high.

This leaves the fast register allocator. The fast register allocator, as
the name suggests, is built for execution speed, not code quality. Thus,
we add special processing to convert the "optimistic" IR into the
"conservative" choice (again at the IR level), which we know it can
handle.

We focus on "rm" for the initial commit, but that can be expanded in the
future for other constraints where Clang generates ++ungood code (like "g").

Fixes: https://github.com/llvm/llvm-project/issues/20571

Cc: Nick Desaulniers <ndesaulniers@google.com>
Cc: Kees Cook <keescook@google.com>
Cc: llvm@lists.linux.dev

bwendling committed:
[Clang][inlineasm] Add special support for "rm" output constraints

Clang isn't able to support multiple constraints on inputs and outputs.
Instead, it picks the "safest" one to use, i.e. the most conseravite. In
the case of "rm" it picks the memory constraint. This leads to obviously
horrible code:

  asm __volatile__ ("pushf\n\t"
                    "popq %0"
                    : "=rm" (x));

is converted to:

	#APP
        pushf
	popq -8(%rsp)
	#NO_APP
	movq	-8(%rsp), %rax

Blech!

This hack^Wchange, makes a special exception for "rm" to use "r" if at
all possible. The "RegMayBeFolded" flag is then used by the register
allocators to allow for the old behavior if register pressure is too
great.

Fixes: https://github.com/llvm/llvm-project/issues/20571

Cc: Nick Desaulniers <ndesaulniers@google.com>
Cc: Kees Cook <keescook@google.com>
Cc: llvm@lists.linux.dev

bwendling commented:
As mentioned in the description, this is a borderline^W hack to get this feature working for x86, which, at least in the Linux kernel, is the only architecture that uses the `"rm"` constraints.

I'm doing this partial fix to get this feature moving, but also I believe that our current support for inline asm is in need of a rewrite. However, given how intertwined inline asm support is with ISel and the register allocators, I'm not even sure it's possible to rewrite it without rewriting major portions of the back end. Even if we wanted to do this, the benefit wouldn't justify the work required.

kees commented on llvm/test/CodeGen/X86/inlineasm-sched-bug.ll:
Hah! Perfect. This improved the codegen for this test. Perfect. :)

kees reviewed (approved):
This looks good to me: the comments are updated where needed; the register preference is limited to x86; "m" is still available on fallback; tests include the different allocators both with and without register pressure.

(Though it looks like a clang format pass is needed.)

arsenm commented on llvm/include/llvm/CodeGen/TargetPassConfig.h:
```diff
@@ -496,6 +496,8 @@ class TargetPassConfig : public ImmutablePass {
 void registerCodeGenCallback(PassInstrumentationCallbacks &PIC,
                              LLVMTargetMachine &);
 
+bool usesGreedyOrDefaultRegisterAllocator();
```
This doesn't work in general and "default" allocator is ambiguous. I'm -1 on anything that tries to expose the allocator in use to clang. I thought the result of the issue thread was fast RA would just start spilling everything?

bwendling replied:
> I thought the result of the issue thread was fast RA would just start spilling everything?

And how do we know that we're in the fast register allocator before we reach the fast register allocator? Also, how does this not work in general?

arsenm replied:
You don't, you just have fastRA do whatever horrible spilling it needs to make it work. Wasn't that what the referenced review in https://github.com/llvm/llvm-project/issues/20571#issuecomment-2061935279 was doing?

qcolombet replied:
+1 on what @arsenm said.
We shouldn't expose the allocator to upper layers.

bwendling replied:
The problem is that code gen from the ISel DAG messes things up enough that "just spilling" the registers isn't easy at all and will require a heinous hack. But if that's what you want, fine.

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Remove function identifying the register allocator used.

bwendling committed:
Merge branch 'llvm:main' into asm-rm

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -4962,6 +4962,11 @@ class TargetLowering : public TargetLoweringBase {
     /// Memory, Other, Unknown.
     TargetLowering::ConstraintType ConstraintType = TargetLowering::C_Unknown;
 
+    /// The register may be folded. This is used if the constraint is "rm",
+    /// where we prefer using a register, but can fall back to a memory slot
+    /// under register pressure.
+    bool MayFoldRegister = false;
+
```
Move to last field? Or could this be expressed as a new RegisterOrMemory ConstraintType?

bwendling replied:
Sure, done.

We already encode the 'MayBeFolded' in the `InlineAsm::Flag` value. If we go with a `C_RegisterOrMemory` type we can reclaim that bit. However, it amounts to the same, and there's not a lot more we can do with it beyond "select one constraint and fall back to the other if the first one fails." I've toyed with the idea of having a completely new way of representing multi-constraints, but it's trickier than it first appears. For instance, we could encode all constraints in the `INLINEASM` machine instruction, selecting the "best" one during register allocation and discarding the rest. This runs into some issues though. First off, cleaning up the unused instructions might be tricky because it relies upon memory analysis, etc. The back end does DCE, but I'm worried that it wouldn't remove *all* instructions.

So I've been going down the rabbit hole that's been suggested in this PR: forcing the fast register allocator to spill all "foldable" registers. As you can imagine, this is easier said than done. There are some situations where the code necessary to support a memory constraint simply doesn't exist and needs to be replicated. Unfortunately, there doesn't appear to be a single way to do that. (If there is and I just haven't found it, please let me know.) So I've resorted to searching for said instructions, and if I don't find them I cobble together the best versions I can and crossing my fingers. This works for several cases, but not for all, as you can imagine.

I'm planning on severely limiting the scope of this feature to "simple" inputs/outputs---no tied constraints, etc.---on X86 and calling it a day. It's a half-step forward, and hacky like nothing else, but the chewing gum, duct tape, and Flex TAPE should hold for the instances we actually care about...

arsenm commented on llvm/test/CodeGen/X86/asm-constraints-rm.ll:
```diff
@@ -0,0 +1,363 @@
+; FAST-I386:    #NO_APP
+entry:
+  %b = getelementptr inbounds i8, ptr %ptr, i64 4
+  %0 = load i32, ptr %b, align 4
```
Use named values in tests

bwendling replied:
I'm not sure what you mean. I think most tests use `%#`, no?

arsenm replied:
No, they mostly use %some_string. opt -passes=instnamer or metarenamer will add names to everything for you

arsenm commented on llvm/test/CodeGen/X86/asm-constraints-rm.ll:
```diff
@@ -0,0 +1,363 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --filter "^\t#" --version 4
+; RUN: llc -mtriple=x86_64-unknown-linux-gnu -O2 -regalloc=greedy < %s | FileCheck --check-prefix=GREEDY-X86_64 %s
+; RUN: llc -mtriple=i386-unknown-linux-gnu -O2 -regalloc=greedy   < %s | FileCheck --check-prefix=GREEDY-I386 %s
+; RUN: llc -mtriple=x86_64-unknown-linux-gnu -O2 -regalloc=basic  < %s | FileCheck --check-prefix=BASIC-X86_64 %s
+; RUN: llc -mtriple=i386-unknown-linux-gnu -O2 -regalloc=basic    < %s | FileCheck --check-prefix=BASIC-I386 %s
+; RUN: llc -mtriple=x86_64-unknown-linux-gnu -O2 -regalloc=fast   < %s | FileCheck --check-prefix=FAST-X86_64 %s
+; RUN: llc -mtriple=i386-unknown-linux-gnu -O2 -regalloc=fast     < %s | FileCheck --check-prefix=FAST-I386 %s
```
Don't need all the -O2, it's the default

bwendling committed:
Run instnamer and remove unneeded '-O2'.

bwendling commented:
@arsenm I appreciate your LGTM, however this version is still a WIP, since I removed the register allocator check that you objected to above.

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Merge branch 'llvm:main' into asm-rm

isanbard commented:
See RFC here:

https://discourse.llvm.org/t/rfc-improving-llvm-support-for-alternative-constraints-in-gnu-extended-inline-assembly/89350

bwendling committed:
Merge branch 'llvm:main' into asm-rm

isanbard committed:
[CodeGen] Add InlineAsmPrepare pass to convert "rm" constraints to "m" for fast regalloc

This patch introduces a new IR-level pass, InlineAsmPrepare, that converts
inline assembly "rm" (register-or-memory) constraints to "m" (memory-only)
constraints when using the fast register allocator on x86 platforms.

Background:
The "rm" constraint allows the compiler to choose between a register or
memory operand. However, LLVM's architecture conservatively picks the most
restrictive interpretation to avoid issues with register pressure. This
causes the fast register allocator to always select registers for "rm"
constraints, even when memory would be more appropriate, leading to
unnecessary spills and suboptimal code.

Solution:
The InlineAsmPrepare pass runs at -O0 before instruction selection and
transforms inline asm calls by:
1. Converting "rm" constraints to "m" (memory-only)
2. Creating allocas for each converted operand
3. For inputs: storing values to allocas and passing pointers
4. For outputs: passing alloca pointers and loading results afterward
5. For tied constraints (e.g., "=rm,0"): storing input to output's alloca
   and passing the same pointer for both the output and tied input

Implementation details:
- InlineAsmPrepare.cpp: New pass that processes CallInst nodes with inline asm
  * Parses constraint strings and identifies "rm" patterns
  * Tracks tied input-output pairs via TiedOutput mapping
  * Reconstructs return values by loading from allocas for converted outputs
  * Properly handles struct returns with mixed converted/unconverted outputs

- SelectionDAG integration:
  * Added MayFoldRegister flag to mark "rm" constraints
  * Modified constraint preference to prefer 'r' over 'm' when MayFoldRegister
    is set, allowing the register allocator to spill to memory if needed
  * Updated AddInlineAsmOperands to propagate the MayFoldRegister flag

- Pass integration:
  * Registered in PassBuilder and PassRegistry
  * Added to TargetPassConfig for -O0 compilation

Test coverage:
Added inline-asm-prepare-memory.ll with three scenarios:
- Input-only: "rm" input → "m" with alloca + store + pointer arg
- Output-only: "=rm" output → "=*m" with alloca + pointer arg + load
- Tied/read-write: "=rm,0" → "=*m,0" with store + dual pointer args + load

The pass only runs on x86 platforms at -O0 to improve code generation for
the fast register allocator without impacting optimized builds.

isanbard committed:
Reformat

isanbard committed:
Use poison instead of undef

isanbard committed:
fix LLVM ABI issues.

isanbard committed:
Correct header order.

bwendling commented:
Some things to mention, which were commented upon above: there's no longer a "usesGreedy...()" predicate, which even I was dubious about using. It turns out that it was easier to just add the pass when optimization wasn't called for. There is still a way to indicate `-O2 -mllvm -regalloc=fast` on the command line, but I think it's perfectly reasonable to throw up our hands and say "it's your problem, dude."

arsenm commented on llvm/lib/CodeGen/InlineAsmPrepare.cpp:
```diff
@@ -0,0 +1,333 @@
+  SmallVector<CallBase *, 4> InlineAsms;
+
+  for_each(F, [&](BasicBlock &BB) {
+    for_each(BB, [&](Instruction &I) {
```
Style guide says to avoid for_each when a regular for will do

bwendling replied:
Done.

isanbard committed:
Follow the style guide re-for loops.

bwendling committed:
Merge branch 'main' into asm-rm

isanbard committed:
Add the appropriate "PreservedAnalyses" data.

Also perform some random code refactoring.

bwendling committed:
Don't make this x86-specific. Also don't completely modify the constraints string, just add an 'indirect' thingy.

bwendling commented:
With this newest commit, I believe that the changes can be reviewed. I've removed the "x86-only" aspect. I've kept the 'InlineAsmPrepare' pass because I think that it can be expanded in the future to improve inline asm even more (there are still incompatibilities between Clang's and GCC's handling of inline asm).

I'm working on comprehensive testing files, which is a bit harder than it appears. But it's coming soon.

PTAL

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Update BPF testcases. These were somehow affected by this new pass, but the pass isn't called during their tests...

bwendling committed:
Update asm call.

yonghong-song commented:
> > ```
> > Failed Tests (2):                                                          
> >   LLVM :: CodeGen/BPF/BTF/func-func-ptr.ll        
> >   LLVM :: CodeGen/BPF/BTF/func-typedef.ll 
> > ```
> >  ...    
> 
> Do you have an idea of what parts of codegen would cause these changes?

In my test, if you do not make changes (i.e. restore to the original test IR) to the above two tests, everything should be okay. For the above two tests, codegen seems having no effect.

bwendling committed:
Revert "Update BPF testcases. These were somehow affected by this new pass, but the pass isn't called during their tests..."

This reverts commit 8d313c1b272ca9433d7e02fa54f4b2a6e75b1b8d.

bwendling committed:
Merge branch 'llvm:main' into asm-rm

h-vetinari commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -2892,13 +2892,19 @@ void CodeGenFunction::EmitAsmStmt(const AsmStmt &S) {
+    // - If this is a register output, then make the inline asm return it
+    //   by-value.
+    // - If this is an "rm" constraint on x86, then treat it like a register
+    //   output. (We'll correct this before ISel if using the FastRA.)
```
>  I've removed the "x86-only" aspect.

Is the "on x86" a left-over here?

bwendling replied:
Left over. Removed. :-)

h-vetinari commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -5979,6 +5980,16 @@ TargetLowering::ParseConstraints(const DataLayout &DL,
+    // register into a spill. This is used for the "rm" constraint, where we
+    // would vastly prefer to use 'r' over 'm', but can't because of LLVM's
+    // architecture picks the most "conservative" constraint to ensure that (in
+    // the case of "rm") register pressure cause bad things to happen.
```
"LLVM's architecture picks [...] constraint to ensure that [X] cause bad things to happen"

This sentence looks mangled; and presumably the architecture tries to _avoid_ bad things from happening 😅 

(another x86 further up the comment)

h-vetinari reviewed (commented):
a comment from the peanut gallery

bwendling committed:
Update comments to remove incorrect info.

bwendling committed:
Add that it's not used for other RAs.

nikic reviewed (commented):
Didn't get far, but a few initial comments. The PR description needs an update

> @nikic I'm still considering merging this with CallBrPrepare, though CallBr now handles more than just ASM instructions. :-) But keeping this separated will help with the code review for this change. In other words, I can merge the two afterwards as an NFC, but before the compiler release. :-)

Not sure I get how that helps with review. I think if you just put your InlineAsmPrepare code into CallBrPrepare things would be a bit simpler because we don't need the pass boilerplate and pipeline test changes anymore? (Doing a CallBrPrepare -> InlineAsmPrepare as an NFC first would be good though.)

bwendling commented on llvm/lib/CodeGen/TargetPassConfig.cpp:
```diff
@@ -985,6 +985,9 @@ void TargetPassConfig::addISelPrepare() {
   if (getOptLevel() != CodeGenOptLevel::None)
     addPass(createObjCARCContractPass());
 
+  if (getOptLevel() == CodeGenOptLevel::None)
```
This is how the fast register allocator is selected (see around line 1173). I'm *not* interested in the value of `OptimizeRegAlloc` (the `-regalloc` flag) because that's hidden for a reason. :-) But remember that this isn't an "optimization" in the normal sense of the word---that the middle-end does its thing. As far as I know, the `OptNone` attribute doesn't change the activities of the back-end at that level (i.e., changing the register allocator on the fly).

bwendling commented on llvm/include/llvm/CodeGen/InlineAsmPrepare.h:
```diff
@@ -0,0 +1,24 @@
+
+class InlineAsmPreparePass : public PassInfoMixin<InlineAsmPreparePass> {
+public:
+  LLVM_ABI PreservedAnalyses run(Function &F, FunctionAnalysisManager &FAM);
```
Weird. `CallBrPrepare` doesn't have that. I'll add it.

bwendling committed:
Add 'isRequired' to the pass class.

bwendling committed:
Spelling.

bwendling commented on llvm/test/CodeGen/X86/asm-constraints-rm.ll:
```diff
@@ -0,0 +1,258 @@
+;
+; FAST-X86_64-LABEL: test1:
+;
+; FAST-I386-LABEL: test1:
```
Yeah. The testcases are in a bad state at the moment. I'm working on comprehensive testcases, so this will be fixed in a short amount of time.

bwendling commented on llvm/test/CodeGen/X86/inline-asm-prepare-memory.ll:
```diff
@@ -0,0 +1,38 @@
+; RUN: llc -O0 -mtriple=x86_64-unknown-linux-gnu -stop-after=inline-asm-prepare < %s | FileCheck %s
```
Roger that!

bwendling committed:
Use better datastructures

bwendling committed:
Merge CallBrPrepare into InlineAsmPrepare, which is the better name for what the pass does.

bwendling commented:
Alright. I merged CallBrPrepare into InlineAsmPrepare. I'm going on to the testcases. PTAL.

(BTW, the whole "pipeline" tests are horribly fragile and annoying.)

bwendling committed:
fixup! Merge CallBrPrepare into InlineAsmPrepare, which is the better name for what the pass does.

Remove dead declaration.

bwendling committed:
Update comments.

bwendling committed:
Restore deleted files

bwendling committed:
Remove temporarily.

bwendling committed:
Rename

bwendling committed:
Merge.

bwendling committed:
Handle indirect output constraints (like =*m) correctly.

bwendling committed:
Update testcase. It's comprehensive now.

bwendling committed:
Use opt and update_test_check.py

bwendling committed:
Merge branch 'main' into asm-rm

bwendling committed:
Merge branch 'llvm:main' into asm-rm

bwendling committed:
Fixed some nits:

- Added checks for the iterator reaching the end.
- Changed some function names to be clearer.
- Added some comments for clarity.

bwendling committed:
Fix small err.

bwendling commented:
> > That approval wasn't rescinded
> 
> I don't think you can reasonably expect a year old approval on a completely different patch to carry over, regardless of whether it was explicitly rescinded (which is an exotic operation on GitHub that few people even know how to perform). I know that one of the reviewers did not intend their approval to carry over (because they specifically complained about the PR reuse going on here) and the other has not even looked at this PR since it pivoted in a different direction (or at least they have given no indication of having looked at it).

I chose to keep the changes in this PR because it kept context. And the reviewers who gave previous approvals did comment on recent changes.

> > and the new implementation was reviewed by many people, including you, in detail.
> 
> My review literally said:
> 
> > Didn't get far, but a few initial comments.
> 
> And I don't see anyone else who has done significant review on this either.
> 
> Which is not surprising, given that this PR was not in a reviewable state until a week ago, when the test coverage was fixed.

The main point of a review is to check the algorithms, code fit, etc. The testcases are obviously important, but any changes to them would be separate from the main code review. The change is large, so I meant for people to start their reviews while I worked on the testcases. Once they were finished, no one had done any further comments, and so I begged for a week for people to comment *anything* on the PR. The very least you could have done is add a short note saying that you'll get to the review soon, or even an emoji, especially because you're an active reviewer. If there is literally zero feedback within a reasonable amount of time, there are only a few inferences one can make: the reviewer has no other comments and is fine with the current approvals, or the reviewer has lost interest.

> > I TOLD YOU THAT I WOULDN'T BE BEST TO COMBINE THE TWO IN THE INITIAL IMPLEMENTATION! However, you and Nick both insisted that it be done this way. There's literally no reason for me to split the two up, because the renaming is NFC. Please don't move goalposts like this.
> 
> I don't think I'm moving goalposts here? I already said back then:
> 
> > Doing a CallBrPrepare -> InlineAsmPrepare as an NFC first would be good though.
> 
> I guess I should have not phrased that as a suggestion?

I must have missed that. My apologies.
